drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+	/* Thread selection */
+	status = CMD_0_THREAD_POS_SET(thread);
+	/* CDMA Mode selection */
+	status |= CMD_0_C_MODE_SET(CT_CDMA_MODE);
+	/* CMD 0 Reg write*/
+	sys_write32(status, (base_address + CDNS_CMD_REG0));
+	return 0;
+}
+
+static struct cdns_cdma_command_descriptor *cdma_allocate_desc_memory(uint32_t page_count,
+								      uint32_t max_page_desc)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+
+	if (page_count > max_page_desc) {
+		page_count = max_page_desc;
+	}
+
+	cdma_desc = k_malloc(sizeof(struct cdns_cdma_command_descriptor) * page_count);
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return NULL;
+	}
+
+	memset(cdma_desc, 0, sizeof(struct cdns_cdma_command_descriptor) * page_count);
+
+	return cdma_desc;
+}
+
+static int cdns_cdma_desc_transfer_finish(struct cadence_nand_params *params,
+					  struct cdns_cdma_command_descriptor *cdma_desc,
+					  uint32_t page_count, uint32_t max_page_desc,
+					  uint32_t ctype, uint32_t cond_start, char *buffer)
+{
+	uint32_t page_count_pass = 0;
+	uint32_t row_address = 0;
+	uint32_t base_address;
+	int ret;
+
+	base_address = params->nand_base;
+
+	while (page_count > 0) {
+		row_address_set(params, &row_address, cond_start);
+
+		if (page_count > max_page_desc) {
+			page_count_pass = max_page_desc;
+			page_count = page_count - max_page_desc;
+			cond_start = cond_start + page_count_pass;
+		} else {
+			page_count_pass = page_count;
+			page_count = page_count - page_count_pass;
+		}
+		for (int index = 0; index < page_count_pass; index++) {
+			cdns_nand_cdma_prepare(NF_TDEF_DEV_NUM, row_address, buffer,
+					       (ctype + index), (page_count_pass - index),
+					       DMA_MS_SEL, VOL_ID, (cdma_desc + index));
+		}
+		ret = cdns_nand_send(base_address, (char *)cdma_desc, NF_TDEF_TRD_NUM);
+
+		if (ret != 0) {
+			k_free(cdma_desc);
+			return ret;
+		}
+
+		if (ctype != CNF_CMD_ERASE) {
+			buffer = buffer + (max_page_desc * params->page_size);
+		}
+
+		ret = cdns_transfer_complete(cdma_desc, params);
+
+		if (ret != 0) {
+			k_free(cdma_desc);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+/**
+ * Perform a CDMA write operation for the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_page_number The starting page number for the write operation.
+ * @param buffer The buffer containing the data to be written.
+ * @param page_count The number of pages to be written.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_cdma_write(struct cadence_nand_params *params, uint32_t start_page_number,
+				char *buffer, uint32_t page_count)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+	int ret;
+
+	cdma_desc = cdma_allocate_desc_memory(page_count, CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT);
+
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
```

Comment:
As far as I can tell, this could be done inside `cdns_cdma_desc_transfer_finish`.
Correct me if I am wrong but the first param to `cdma_allocate_desc_memory` is always third param to `cdns_cdma_desc_transfer_finish` and second param to the former is always fourth param to the later.
Wouldn't it save flash to move the block above to the `cdns_cdma_desc_transfer_finish` ?

Reply:
moved cdma_allocate_desc_memory functionality to cdns_cdma_desc_transfer_finish block.